### PR TITLE
feat: Add PubSub Driver 

### PR DESCRIPTION
**Pull request by kangoo13**

### CODE REVIEW

File: src/PubSub/Queue.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    /**
+     * Get the size of the queue.
+     * PubSubClient have no method to retrieve the size of the queue.
+     * To be updated if the API allow to get that data.
+     *
+     * @param  string $queue
+     * @return int
+     */
+    public function size($queue = null)
+    {
```
Comment (fdv) [inline src/PubSub/Queue.php:50]: Huh????

File: src/PubSub/Queue.php [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+    /**
+     * Create a new subscription to a topic.
+     *
+     * @param  \Google\Cloud\PubSub\Topic $topic
+     * @return \Google\Cloud\PubSub\Subscription
+     */
+    public function subscribeToTopic(Topic $topic)
+    {
+        $subscription = $topic->subscription($this->getSubscriberName($topic));
```
Comment (fdv) [inline src/PubSub/Queue.php]: `return ($topic->subscription($this->getSubscriberName($topic))); ?`

File: src/PubSub/Queue.php [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    /**
+     * Get the current topic
+     *
+     * @param $queue
+     * @return Topic
+     */
+    public function getTopic($queue)
+    {
+        $queue = $queue ?: $this->default;
+        $topic = $this->pubsub->topic($queue);
```
Comment (fdv) [inline src/PubSub/Queue.php]: `return ($this->pubsub->topic($queue));` ?

AKA why initializing a var and adding memory to the heap only to return it?

File: src/PubSub/Queue.php [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+     * Acknowledge a message.
+     *
+     * @param  \Google\Cloud\PubSub\Message $message
+     * @param  string                       $queue
+     */
+    public function acknowledge(Message $message, $queue = null)
+    {
+        $topic        = $this->getTopic($queue);
+        $subscription = $topic->subscription($this->getSubscriberName($topic));
+        $subscription->acknowledge($message);
```
Comment (fdv) [inline src/PubSub/Queue.php]: `$topic->subscription($this->getSubscriberName($topic))->acknowledge($message);`

AKA "memory might be cheap, it's not that cheap"

File: src/PubSub/Queue.php [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+     *
+     * @param  string $payload
+     * @param  string $queue
+     * @param  array  $options
+     * @return array
+     */
+    public function pushRaw($payload, $queue = null, array $options = null)
+    {
+        $topic   = $this->getTopic($queue);
+        $publish = $topic->publish([
```
Comment (fdv) [inline src/PubSub/Queue.php]: You can `return ($topic->publish([
            'data'       => $payload,
            'attributes' => $options,
        ]));` as you only initialize `$publish` to return it. It means you allocate memory on the heap you will never use anyway and this also means garbage collection will have more work in the future. Only initialize a variable for a reason.

**Review by fdv — CHANGES_REQUESTED** [review]

File: src/PubSub/Queue.php (hunk 1 above)
Comment (kangoo13) [reply src/PubSub/Queue.php:50]: We must implement the interface Illuminate\Contracts\Queue\Queue, and we can't know how much messages are left in a PubSub queue

**Review by fdv — APPROVED** [review]
👍

**Pull request merged**